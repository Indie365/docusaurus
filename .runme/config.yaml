version: 1.0
publish: app
services:
  app:
    build:
      type: dockerfile
      config: ./.runme/Dockerfile.runme
    ports:
      - container: 3000
        public: 80