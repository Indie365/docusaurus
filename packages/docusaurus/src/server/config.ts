/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import fs from 'fs-extra';
import logger from '@docusaurus/logger';
import {
  DEFAULT_CONFIG_FILE_NAME,
  DEFAULT_THEME_FILE_NAME,
  findAsyncSequential,
  loadFreshModule,
} from '@docusaurus/utils';
import {validateConfig} from './configValidation';
import type {DocusaurusConfig, LoadContext} from '@docusaurus/types';

async function getConventionalSiteConfigPath(siteDir: string) {
  // We could support .mjs, .ts, etc. in the future
  const candidates = ['.ts', '.mts', '.cts', '.js', '.mjs', '.cjs'].map(
    (ext) => DEFAULT_CONFIG_FILE_NAME + ext,
  );
  const configPath = await findAsyncSequential(
    candidates.map((file) => path.join(siteDir, file)),
    fs.pathExists,
  );
  if (!configPath) {
    const errorMessage = logger.interpolate`No config file found.
Expected one of:${candidates}
You can provide a custom config path with the code=${'--config'} option.`;
    throw new Error(errorMessage);
  }
  return configPath;
}

export async function loadSiteConfig({
  siteDir,
  customConfigFilePath,
}: {
  siteDir: string;
  customConfigFilePath?: string;
}): Promise<Pick<LoadContext, 'siteConfig' | 'siteConfigPath'>> {
  const siteConfigPath = customConfigFilePath
    ? path.resolve(siteDir, customConfigFilePath)
    : await getConventionalSiteConfigPath(siteDir);

  if (!(await fs.pathExists(siteConfigPath))) {
    throw new Error(`Config file at "${siteConfigPath}" not found.`);
  }

  const importedConfig = await loadFreshModule(siteConfigPath);

  const loadedConfig: unknown =
    typeof importedConfig === 'function'
      ? await importedConfig()
      : await importedConfig;

  const siteConfig = validateConfig(
    loadedConfig,
    path.relative(siteDir, siteConfigPath),
  );
  return {siteConfig, siteConfigPath};
}

async function findConventionalThemeConfigPath(
  siteDir: string,
): Promise<string | undefined> {
  // We could support .mjs, .ts, etc. in the future
  const candidates = ['.tsx', '.ts', '.jsx', '.js'].map(
    (ext) => DEFAULT_THEME_FILE_NAME + ext,
  );
  const themeConfigPath = await findAsyncSequential(
    candidates.map((file) => path.join(siteDir, file)),
    fs.pathExists,
  );

  return themeConfigPath;
}

// TODO add tests for this
export async function findThemeConfigPath(
  siteDir: string,
  siteConfig: DocusaurusConfig,
): Promise<string | undefined> {
  // TODO add support for custom themeConfig file path
  // EX: siteConfig.themeConfig: './theme.tsx'
  // Note: maybe it would be simpler to provide this path through the CLI?
  if (typeof siteConfig.themeConfig === 'string') {
    const customThemeConfigPath = siteConfig.themeConfig;
    if (!(await fs.pathExists(customThemeConfigPath))) {
      throw new Error(
        `Theme config file at "${customThemeConfigPath}" not found.`,
      );
    }
    return customThemeConfigPath;
  }
  const conventionalThemeConfigPath = await findConventionalThemeConfigPath(
    siteDir,
  );
  // In Docusaurus v3 we require users to provide either the theme config
  // through the conventional path, or through the legacy siteConfig attribute
  // To avoid issues we prevent users to not provide any theme config at all
  if (
    !conventionalThemeConfigPath &&
    Object.keys(siteConfig.themeConfig ?? {}).length === 0
  ) {
    throw new Error(
      `Theme config file couldn't be found at ${DEFAULT_THEME_FILE_NAME}.js or ${DEFAULT_THEME_FILE_NAME}.tsx`,
    );
  }
  return conventionalThemeConfigPath;
}
