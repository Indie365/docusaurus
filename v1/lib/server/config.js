/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const fs = require('fs-extra');

module.exports = function loadConfig(configPath, deleteCache = true) {
  if (deleteCache) {
    delete require.cache[configPath];
  }
  let config = {};
  if (fs.existsSync(configPath)) {
    config = require(configPath); // eslint-disable-line
  }

  /* docsUrl */
  if (config.docsUrl && (config.docsUrl === '' || config.docsUrl === '/')) {
    config.docsUrl = '';
  } else if (!config.docsUrl) {
    config.docsUrl = 'docs';
  }

  return config;
};
