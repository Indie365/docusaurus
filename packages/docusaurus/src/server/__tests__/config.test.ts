/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {loadConfig} from '../config';

describe('loadConfig', () => {
  test('website with valid siteConfig', async () => {
    const fixtures = path.join(__dirname, '__fixtures__');
    const siteDir = path.join(fixtures, 'simple-site');
    const config = loadConfig(siteDir);
    expect(config).toMatchInlineSnapshot(`
      Object {
        "baseUrl": "/",
        "favicon": "img/docusaurus.ico",
        "organizationName": "endiliey",
        "plugins": Array [
          Object {
            "name": "@docusaurus/plugin-content-docs",
            "options": Object {
              "path": "../docs",
              "sidebarPath": "/mnt/c/Users/endij/Desktop/Linux/Docusaurus/packages/docusaurus/src/server/__tests__/__fixtures__/simple-site/sidebars.json",
            },
          },
          Object {
            "name": "@docusaurus/plugin-content-pages",
          },
        ],
        "projectName": "hello",
        "tagline": "Hello World",
        "title": "Hello",
        "url": "https://docusaurus.io",
      }
    `);
    expect(config).not.toEqual({});
  });

  test('website with incomplete siteConfig', () => {
    const siteDir = path.join(__dirname, '__fixtures__', 'bad-site');
    expect(() => {
      loadConfig(siteDir);
    }).toThrowErrorMatchingInlineSnapshot(
      `"The required field(s) 'favicon', 'tagline', 'url' are missing from docusaurus.config.js"`,
    );
  });

  test('website with useless field (wrong field) in siteConfig', () => {
    const siteDir = path.join(__dirname, '__fixtures__', 'wrong-site');
    expect(() => {
      loadConfig(siteDir);
    }).toThrowErrorMatchingInlineSnapshot(
      `"The required field(s) 'favicon' are missing from docusaurus.config.js"`,
    );
  });

  test('website with no siteConfig', () => {
    const siteDir = path.join(__dirname, '__fixtures__', 'nonExisting');
    expect(() => {
      loadConfig(siteDir);
    }).toThrowErrorMatchingInlineSnapshot(`"docusaurus.config.js not found"`);
  });
});
