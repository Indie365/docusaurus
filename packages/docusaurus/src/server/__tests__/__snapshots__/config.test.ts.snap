// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadConfig website with incomplete siteConfig 1`] = `"\\"favicon\\" is required"`;

exports[`loadConfig website with no siteConfig 1`] = `"docusaurus.config.js not found"`;

exports[`loadConfig website with useless field (wrong field) in siteConfig 1`] = `"\\"favicon\\" is required"`;

exports[`loadConfig website with valid siteConfig 1`] = `
Object {
  "baseUrl": "/",
  "customFields": Object {},
  "favicon": "img/docusaurus.ico",
  "organizationName": "endiliey",
  "plugins": Array [
    Array [
      "@docusaurus/plugin-content-docs",
      Object {
        "path": "../docs",
      },
    ],
    "@docusaurus/plugin-content-pages",
  ],
  "projectName": "hello",
  "tagline": "Hello World",
  "themeConfig": Object {},
  "themes": Array [],
  "title": "Hello",
  "url": "https://docusaurus.io",
}
`;
